daml-foundations/daml-ghc/ide/src/Development/IDE/State/Rules/Daml.hs

```diff
@@ -258,6 +258,29 @@ encodeModule lfVersion m =
         | isAbsolute file -> use_ EncodeModule file
       _ -> pure $ SS.encodeModule lfVersion m
 
+getScenarioRootsRule :: Rules ()
+getScenarioRootsRule =
+    defineNoFile $ \GetScenarioRoots -> do
+        alwaysRerun
+        Env{..} <- getServiceEnv
+        DamlEnv{..} <- getDamlServiceEnv
+        filesOfInterest <- liftIO $ readVar envOfInterestVar
+        openVRs <- liftIO $ readVar envOpenVirtualResources
+        let files = Set.toList (filesOfInterest `Set.union` Set.map vrScenarioFile openVRs)
+        deps <- forP files $ \file -> do
+            transitiveDeps <- maybe [] transitiveModuleDeps <$> use GetDependencies file
+            pure $ Map.fromList [ (f, file) | f <- transitiveDeps ]
+        -- We want to ensure that files of interest always map to themselves even if there are dependencies
+        -- between files of interest so we union them separately. (`Map.union` is left-biased.)
+        pure $ Map.fromList [(f,f) | f <- files] `Map.union` Map.unions deps
+
+getScenarioRootRule :: Rules ()
+getScenarioRootRule =
+    defineEarlyCutoff $ \GetScenarioRoot file -> do
+        ctxRoots <- use_ GetScenarioRoots ""
+        case Map.lookup file ctxRoots of
+            Nothing -> liftIO $ throwIO $ ScenarioRootException file
```

Comment:
If this is an internal exception just call `fail` or similar.

Reply:
Changed to `fail`